kubernetes-client/src/Kubernetes/Client/Config.hs

```diff
@@ -1,136 +1,158 @@
-import           System.X509                (getSystemCertificateStore)
+{-# LANGUAGE OverloadedStrings #-}
+
+module Kubernetes.Client.Config
+  ( KubeConfigSource(..)
+  , addCACertData
+  , addCACertFile
+  , applyAuthSettings
+  , clientHooksL
+  , Kubernetes.Client.Config.cluster
+  , defaultTLSClientParams
+  , disableServerCertValidation
+  , disableServerNameValidation
+  , disableValidateAuthMethods
+  , kubeClient
+  , loadPEMCerts
+  , newManager
+  , onCertificateRequestL
+  , onServerCertificateL
+  , parsePEMCerts
+  , serviceAccountDir
+  , setCAStore
+  , setClientCert
+  , setMasterURI
+  , setTokenAuth
+  , tlsValidation
+  )
+where
+
+import qualified Kubernetes.OpenAPI.Core as K
+
+import           Control.Applicative                 ((<|>))
+import           Control.Exception.Safe              (MonadThrow, throwM)
+import           Control.Monad.IO.Class              (MonadIO, liftIO)
+import qualified Data.ByteString                     as B
+import qualified Data.ByteString.Base64              as B64
+import qualified Data.ByteString.Lazy                as LazyB
+import           Data.Either.Combinators
+import           Data.Function                       ((&))
+import           Data.Maybe
+import qualified Data.Text                           as T
+import qualified Data.Text.Encoding                  as T
+import qualified Data.Text.IO                        as T
+import           Data.Yaml
+import           Kubernetes.Client.Auth.ClientCert
+import           Kubernetes.Client.Auth.GCP
+import           Kubernetes.Client.Auth.OIDC
+import           Kubernetes.Client.Auth.Token
+import           Kubernetes.Client.Internal.TLSUtils
+import           Kubernetes.Client.KubeConfig
+import           Network.Connection                  (TLSSettings (..))
+import qualified Network.HTTP.Client                 as NH
+import           Network.HTTP.Client.TLS             (mkManagerSettings)
+import qualified Network.TLS                         as TLS
+import           System.Environment                  (getEnv)
+import           System.FilePath
+
+data KubeConfigSource = KubeConfigFile FilePath
+                      | KubeConfigCluster
+
+{-|
+  Creates 'NH.Manager' and 'K.KubernetesClientConfig' for a given
+  'KubeConfigSource'. It is recommended that multiple 'kubeClient' invocations
+  across an application share an 'OIDCCache', this makes sure updation of OAuth
+  token is synchronized across all the different clients being used.
+-}
+kubeClient
+  :: OIDCCache
+  -> KubeConfigSource
+  -> IO (NH.Manager, K.KubernetesClientConfig)
+kubeClient oidcCache (KubeConfigFile f) = do
+  kubeConfigFile <- decodeFileThrow f
+  uri <- getCluster kubeConfigFile
+         & fmap server
+         & either (const $ pure "localhost:8080") return
+  t <- defaultTLSClientParams
+       & fmap (tlsValidation kubeConfigFile)
+       & (>>= (addCACertData kubeConfigFile))
+       & (>>= addCACertFile kubeConfigFile (takeDirectory f))
```

Comment:
Some of this partial application makes it hard to follow what's going on here, would it be possible to break some of this stuff out into the main `do` block and to reduce some of the combinator usage?

Reply:
I have extracted this out into a function and given each step a name so it is more followable.
I found was doing similar things in one more place, I have fixed that too.